// @better-auth/infra/dist/email.mjs
import { i as INFRA_USER_AGENT, n as INFRA_API_URL } from "./constants-CLYqEwMV.mjs";
import { logger } from "better-auth";
import { env } from "@better-auth/core/env";
import { createFetch } from "@better-fetch/fetch";
//#region src/email.ts
/**
* Email sending module for @better-auth/infra
*
* This module provides email sending functionality that integrates with
* Better Auth Infra's template system.
*/
/**
* Email template definitions with their required variables
*/
const EMAIL_TEMPLATES = {
	"verify-email": { variables: {} },
	"reset-password": { variables: {} },
	"change-email": { variables: {} },
	"sign-in-otp": { variables: {} },
	"verify-email-otp": { variables: {} },
	"reset-password-otp": { variables: {} },
	"magic-link": { variables: {} },
	"two-factor": { variables: {} },
	invitation: { variables: {} },
	"application-invite": { variables: {} },
	"delete-account": { variables: {} },
	"stale-account-user": { variables: {} },
	"stale-account-admin": { variables: {} }
};
/**
* Create an email sender instance
*/
function createEmailSender(config) {
	const baseUrl = config?.apiUrl || INFRA_API_URL;
	const apiUrl = baseUrl.endsWith("/api") ? baseUrl : `${baseUrl}/api`;
	const apiKey = config?.apiKey || env.BETTER_AUTH_API_KEY || "";
	if (!apiKey) logger.warn("[Dash] No API key provided for email sending. Set BETTER_AUTH_API_KEY environment variable or pass apiKey in config.");
	const $api = createFetch({
		baseURL: apiUrl,
		headers: {
			"user-agent": INFRA_USER_AGENT,
			Authorization: `Bearer ${apiKey}`
		},
		timeout: config?.apiTimeout ?? 3e3
	});
	/**
	* Send an email using a template from Better Auth Infra
	*/
	async function send(options) {
		if (!apiKey) return {
			success: false,
			error: "API key not configured"
		};
		try {
			const { data, error } = await $api("/v1/email/send", {
				method: "POST",
				body: {
					template: options.template,
					to: options.to,
					variables: options.variables || {},
					subject: options.subject
				}
			});
			if (error) return {
				success: false,
				error: error.message || `HTTP ${error.status}`
			};
			if (typeof data !== "object" || Array.isArray(data)) return {
				success: false,
				error: "Failed to parse JSON"
			};
			return {
				success: true,
				messageId: data.messageId
			};
		} catch (error) {
			logger.warn("[Dash] Email send failed:", error);
			return {
				success: false,
				error: error instanceof Error ? error.message : "Failed to send email"
			};
		}
	}
	/**
	* Send bulk emails using a template from Better Auth Infra
	*/
	async function sendBulk(options) {
		if (!apiKey) return {
			success: false,
			failures: Object.fromEntries(options.emails.map((e) => [e.to, [{ error: "API key not configured" }]]))
		};
		try {
			const { data, error } = await $api("/v1/email/send-bulk", {
				method: "POST",
				body: {
					template: options.template,
					emails: options.emails.map((e) => ({
						to: e.to,
						variables: e.variables || {}
					})),
					subject: options.subject,
					variables: options.variables || {}
				}
			});
			if (error) return {
				success: false,
				failures: Object.fromEntries(options.emails.map((e) => [e.to, [{ error: error.message || `HTTP ${error.status}` }]]))
			};
			if (typeof data !== "object" || Array.isArray(data) || typeof data.success !== "boolean") return {
				success: false,
				failures: Object.fromEntries(options.emails.map((e) => [e.to, [{ error: "Failed to parse JSON" }]]))
			};
			return {
				success: data.success,
				failures: data.failures
			};
		} catch (error) {
			logger.warn("[Dash] Bulk email send failed:", error);
			return {
				success: false,
				failures: Object.fromEntries(options.emails.map((e) => [e.to, [{ error: error instanceof Error ? error.message : "Failed to send bulk emails" }]]))
			};
		}
	}
	/**
	* Get available email templates
	*/
	async function getTemplates() {
		if (!apiKey) return [];
		try {
			const { data, error } = await $api("/v1/email/templates", { method: "GET" });
			if (error || !data || !Array.isArray(data)) return [];
			return data;
		} catch (error) {
			logger.warn("[Dash] Failed to fetch email templates:", error);
			return [];
		}
	}
	return {
		send,
		sendBulk,
		getTemplates
	};
}
/**
* Send an email using the Better Auth dashboard's email templates.
*
* @example
* ```ts
* import { sendEmail } from "@better-auth/infra";
*
* // Type-safe - variables are inferred from template
* await sendEmail({
*   template: "reset-password",
*   to: "user@example.com",
*   variables: {
*     resetLink: "https://...",
*     userEmail: "user@example.com",
*   },
* });
* ```
*/
async function sendEmail(options, config) {
	return createEmailSender(config).send(options);
}
/**
* Send bulk emails using the Better Auth dashboard's email templates.
*
* @example
* ```ts
* import { sendBulkEmails } from "@better-auth/infra";
*
* const result = await sendBulkEmails({
*   template: "reset-password",
*   emails: [
*     { to: "user1@example.com", variables: { resetLink: "...", userEmail: "user1@example.com" } },
*     { to: "user2@example.com", variables: { resetLink: "...", userEmail: "user2@example.com" } },
*   ],
* });
* ```
*/
async function sendBulkEmails(options, config) {
	return createEmailSender(config).sendBulk(options);
}
//#endregion
export { EMAIL_TEMPLATES, createEmailSender, sendBulkEmails, sendEmail };
