// @better-auth/infra/dist/pow-retry-D2RRftJr.mjs
//#region src/dash-client.ts
function resolveDashUserId(input, options) {
	return input.userId || options?.resolveUserId?.({
		userId: input.userId,
		user: input.user,
		session: input.session
	}) || input.user?.id || input.session?.user?.id || void 0;
}
const dashClient = (options) => {
	return {
		id: "dash",
		getActions: ($fetch) => ({ dash: {
			getAuditLogs: async (input = {}) => {
				const userId = resolveDashUserId(input, options);
				return $fetch("/events/audit-logs", {
					method: "GET",
					query: {
						limit: input.limit,
						offset: input.offset,
						organizationId: input.organizationId,
						identifier: input.identifier,
						eventType: input.eventType,
						userId
					}
				});
			},
			getAllAuditLogs: async (input = {}) => {
				return $fetch("/events/all-audit-logs", {
					method: "GET",
					query: {
						limit: input.limit,
						offset: input.offset,
						organizationId: input.organizationId,
						userId: input.userId,
						eventType: input.eventType,
						identifier: input.identifier
					}
				});
			}
		} }),
		pathMethods: {
			"/events/audit-logs": "GET",
			"/events/all-audit-logs": "GET"
		}
	};
};
//#endregion
//#region src/crypto.ts
function randomBytes(length) {
	const bytes = new Uint8Array(length);
	crypto.getRandomValues(bytes);
	return bytes;
}
function bytesToHex(bytes) {
	return Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("");
}
async function hash(message) {
	const msgBuffer = new TextEncoder().encode(message);
	const hashBuffer = await crypto.subtle.digest("SHA-256", msgBuffer);
	return bytesToHex(new Uint8Array(hashBuffer));
}
//#endregion
//#region src/identification-client.ts
function generateRequestId() {
	const hex = bytesToHex(randomBytes(16));
	return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20, 32)}`;
}
async function identify($kv, payload) {
	await $kv("/identify", {
		method: "POST",
		body: payload
	});
}
//#endregion
//#region src/sentinel/identify-url.ts
const DEFAULT_IDENTIFY_URL = "https://kv.better-auth.com";
function normalizeIdentifyUrl(url) {
	return url.replace(/\/$/, "");
}
/** True when the URL targets project-scoped ingestion, e.g. .../projects/{orgId}. */
function isProjectScopedIdentifyUrl(url) {
	return /\/projects\/[^/]+$/.test(normalizeIdentifyUrl(url));
}
const DEFAULT_GLOBAL_INGESTION_WARNING = "[Sentinel] Default global identify ingestion is active but not recommended. Get your ingestion url from your project settings page (e.g. https://kv.better-auth.com/projects/{id}) and configure sentinelClient({ identifyUrl }) with it.";
const CONFIGURED_GLOBAL_INGESTION_WARNING = "[Sentinel] Global identify ingestion is configured but not recommended. Get your ingestion url from your project settings page (e.g. https://kv.better-auth.com/projects/{id}) and configure sentinelClient({ identifyUrl }) with it.";
function warnOnGlobalIdentifyIngestion(url, message) {
	if (isProjectScopedIdentifyUrl(url)) return;
	console.warn(message);
}
function resolveSentinelClientIdentifyUrl(options) {
	const explicit = options?.identifyUrl?.trim();
	if (explicit) {
		const url = normalizeIdentifyUrl(explicit);
		warnOnGlobalIdentifyIngestion(url, CONFIGURED_GLOBAL_INGESTION_WARNING);
		return url;
	}
	const fromEnv = options?.envKvUrl?.trim();
	if (fromEnv) {
		const url = normalizeIdentifyUrl(fromEnv);
		warnOnGlobalIdentifyIngestion(url, CONFIGURED_GLOBAL_INGESTION_WARNING);
		return url;
	}
	const url = DEFAULT_IDENTIFY_URL;
	warnOnGlobalIdentifyIngestion(url, DEFAULT_GLOBAL_INGESTION_WARNING);
	return url;
}
//#endregion
//#region src/sentinel/pow.ts
function hasLeadingZeroBits(hash, bits) {
	const fullHexChars = Math.floor(bits / 4);
	const remainingBits = bits % 4;
	for (let i = 0; i < fullHexChars; i++) if (hash[i] !== "0") return false;
	if (remainingBits > 0 && fullHexChars < hash.length) {
		if (parseInt(hash[fullHexChars], 16) > (1 << 4 - remainingBits) - 1) return false;
	}
	return true;
}
async function solvePoWChallenge(challenge) {
	const { nonce, difficulty } = challenge;
	let counter = 0;
	while (true) {
		if (hasLeadingZeroBits(await hash(`${nonce}:${counter}`), difficulty)) return {
			nonce,
			counter
		};
		counter++;
		if (counter % 1e3 === 0) await new Promise((resolve) => setTimeout(resolve, 0));
		if (counter > 1e8) throw new Error("PoW challenge took too long to solve");
	}
}
function decodeBase64ToUtf8(encoded) {
	if (typeof globalThis.atob === "function") return globalThis.atob(encoded);
	throw new Error("[Sentinel] Base64 decode requires atob (browser, Hermes, or Bun)");
}
function encodeUtf8ToBase64(str) {
	if (typeof globalThis.btoa === "function") return globalThis.btoa(str);
	const bytes = new TextEncoder().encode(str);
	const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
	let out = "";
	for (let i = 0; i < bytes.length; i += 3) {
		const b0 = bytes[i];
		const b1 = bytes[i + 1] ?? 0;
		const b2 = bytes[i + 2] ?? 0;
		const triple = b0 << 16 | b1 << 8 | b2;
		const pad = i + 2 >= bytes.length ? i + 1 >= bytes.length ? 2 : 1 : 0;
		out += chars[triple >> 18 & 63];
		out += chars[triple >> 12 & 63];
		out += pad < 2 ? chars[triple >> 6 & 63] : "=";
		out += pad < 1 ? chars[triple & 63] : "=";
	}
	return out;
}
function decodePoWChallenge(encoded) {
	try {
		const decoded = decodeBase64ToUtf8(encoded);
		return JSON.parse(decoded);
	} catch {
		return null;
	}
}
function encodePoWSolution(solution) {
	return encodeUtf8ToBase64(JSON.stringify(solution));
}
//#endregion
//#region src/sentinel/pow-retry.ts
/**
* better-fetch clears its internal timeout once the first response is received.
* For the PoW retry we apply the same `timeout` value again so the second leg
* cannot exceed the client's configured limit.
*/
function createPowRetryTimeout(timeoutMs) {
	if (typeof timeoutMs !== "number" || timeoutMs <= 0) return {
		signal: void 0,
		cleanup: void 0
	};
	const controller = new AbortController();
	const id = setTimeout(() => controller.abort(), timeoutMs);
	return {
		signal: controller.signal,
		cleanup: () => clearTimeout(id)
	};
}
//#endregion
export { resolveSentinelClientIdentifyUrl as a, bytesToHex as c, solvePoWChallenge as i, hash as l, decodePoWChallenge as n, generateRequestId as o, encodePoWSolution as r, identify as s, createPowRetryTimeout as t, dashClient as u };
